Index: Bootstrap-v1.16/crt0_gnu.S
===================================================================
--- Bootstrap-v1.16.orig/crt0_gnu.S	2011-10-11 17:19:04.091131234 +0200
+++ Bootstrap-v1.16/crt0_gnu.S	2011-10-11 17:19:29.865169000 +0200
@@ -56,7 +56,7 @@
 	b 	swi_vector   	/* Software Interrupt */
 	b 	pabt_vector  	/* Prefetch Abort */
 	b 	dabt_vector  	/* Data Abort */
-.word		_edata		/* Size of the image for SAM-BA */
+.word		_edata - reset		/* Size of the image for SAM-BA */
 	b 	irq_vector	/* IRQ : read the AIC */
 	b 	fiq_vector      /* FIQ */
 
