Index: Bootstrap-v1.16/include/nand_ids.h
===================================================================
--- Bootstrap-v1.16.orig/include/nand_ids.h	2016-08-05 13:16:52.623879900 +0200
+++ Bootstrap-v1.16/include/nand_ids.h	2016-08-05 13:17:12.541006300 +0200
@@ -46,6 +46,7 @@
 	{0x20aa, 0x800, 0x20000, 0x800, 0x40, 0x0, "STMNAND02GR3B\0"},	
 	{0x20da, 0x800, 0x20000, 0x800, 0x40, 0x0, "STMNAND02GW3B\0"},
 	{0x2caa, 0x800, 0x20000, 0x800, 0x40, 0x0, "MT29F2G08ABD\0"},
+	{0xc2da, 0x800, 0x20000, 0x800, 0x40, 0x0, "MX30LF2GE8AB\0"},
 	{0,}
 };
 
