Index: linux-2.6.39/arch/arm/mach-at91/include/mach/board.h
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/include/mach/board.h
+++ linux-2.6.39/arch/arm/mach-at91/include/mach/board.h
@@ -112,6 +112,7 @@ struct atmel_nand_data {
 	u8		ale;		/* address line number connected to ALE */
 	u8		cle;		/* address line number connected to CLE */
 	u8		bus_width_16;	/* buswidth is 16 bit */
+	u8		use_dma;	/* enable / disable DMA for buffer transfer */
 	struct mtd_partition* (*partition_info)(int, int*);
 };
 extern void __init at91_add_device_nand(struct atmel_nand_data *data);
Index: linux-2.6.39/drivers/mtd/nand/atmel_nand.c
===================================================================
--- linux-2.6.39.orig/drivers/mtd/nand/atmel_nand.c
+++ linux-2.6.39/drivers/mtd/nand/atmel_nand.c
@@ -539,6 +539,8 @@ static int __init atmel_nand_probe(struc
 	if (host->board->rdy_pin)
 		nand_chip->dev_ready = atmel_nand_device_ready;
 
+	use_dma = host->board->use_dma;
+
 	regs = platform_get_resource(pdev, IORESOURCE_MEM, 1);
 	if (!regs && hard_ecc) {
 		printk(KERN_ERR "atmel_nand: can't get I/O resource "
