Index: linux-2.6.39/include/linux/usb/atmel_usba_udc.h
===================================================================
--- linux-2.6.39.orig/include/linux/usb/atmel_usba_udc.h
+++ linux-2.6.39/include/linux/usb/atmel_usba_udc.h
@@ -17,6 +17,7 @@ struct usba_platform_data {
 	int			vbus_pin;
 	int			vbus_pin_inverted;
 	int			num_ep;
+	int			force_full_speed;
 	struct usba_ep_data	ep[0];
 };
 
Index: linux-2.6.39/drivers/usb/gadget/atmel_usba_udc.c
===================================================================
--- linux-2.6.39.orig/drivers/usb/gadget/atmel_usba_udc.c
+++ linux-2.6.39/drivers/usb/gadget/atmel_usba_udc.c
@@ -1148,12 +1148,18 @@ static int do_test_mode(struct usba_udc
 	switch (test_mode) {
 	case 0x0100:
 		/* Test_J */
-		usba_writel(udc, TST, USBA_TST_J_MODE);
+		if(udc->force_full_speed)
+			usba_writel(udc, TST, USBA_TST_J_MODE | USBA_BF(SPEED_CFG, USBA_SPEED_CFG_FORCE_FULL));
+		else
+			usba_writel(udc, TST, USBA_TST_J_MODE);
 		dev_info(dev, "Entering Test_J mode...\n");
 		break;
 	case 0x0200:
 		/* Test_K */
-		usba_writel(udc, TST, USBA_TST_K_MODE);
+		if(udc->force_full_speed)
+			usba_writel(udc, TST, USBA_TST_K_MODE | USBA_BF(SPEED_CFG, USBA_SPEED_CFG_FORCE_FULL));
+		else
+			usba_writel(udc, TST, USBA_TST_K_MODE);
 		dev_info(dev, "Entering Test_K mode...\n");
 		break;
 	case 0x0300:
@@ -1190,7 +1196,10 @@ static int do_test_mode(struct usba_udc
 			dev_err(dev, "Test_Packet: ep0 not mapped\n");
 		} else {
 			usba_ep_writel(ep, CTL_ENB, USBA_EPT_ENABLE);
-			usba_writel(udc, TST, USBA_TST_PKT_MODE);
+			if(udc->force_full_speed)
+				usba_writel(udc, TST, USBA_TST_PKT_MODE | USBA_BF(SPEED_CFG, USBA_SPEED_CFG_FORCE_FULL));
+			else
+				usba_writel(udc, TST, USBA_TST_PKT_MODE);
 			memcpy_toio(ep->fifo, test_packet_buffer,
 					sizeof(test_packet_buffer));
 			usba_ep_writel(ep, SET_STA, USBA_TX_PK_RDY);
@@ -2021,6 +2030,13 @@ static int __init usba_udc_probe(struct
 		}
 	}
 
+	udc->force_full_speed = pdata->force_full_speed;
+	//udc->force_full_speed = 1;
+
+	if(udc->force_full_speed)
+		usba_writel(udc, TST,
+				USBA_BF(SPEED_CFG, USBA_SPEED_CFG_FORCE_FULL));
+
 	usba_init_debugfs(udc);
 	for (i = 1; i < pdata->num_ep; i++)
 		usba_ep_init_debugfs(udc, &usba_ep[i]);
Index: linux-2.6.39/drivers/usb/gadget/atmel_usba_udc.h
===================================================================
--- linux-2.6.39.orig/drivers/usb/gadget/atmel_usba_udc.h
+++ linux-2.6.39/drivers/usb/gadget/atmel_usba_udc.h
@@ -79,6 +79,7 @@
 /* Bitfields in USBA_TST */
 #define USBA_SPEED_CFG_OFFSET			0
 #define USBA_SPEED_CFG_SIZE			2
+#define USBA_SPEED_CFG_NORMAL		0
 #define USBA_TST_J_MODE				(1 <<  2)
 #define USBA_TST_K_MODE				(1 <<  3)
 #define USBA_TST_PKT_MODE			(1 <<  4)
@@ -336,6 +337,8 @@ struct usba_udc {
 	struct dentry *debugfs_root;
 	struct dentry *debugfs_regs;
 #endif
+
+	int force_full_speed;
 };
 
 static inline struct usba_ep *to_usba_ep(struct usb_ep *ep)
Index: linux-2.6.39/arch/arm/mach-at91/at91sam9g45_devices.c
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/at91sam9g45_devices.c
+++ linux-2.6.39/arch/arm/mach-at91/at91sam9g45_devices.c
@@ -263,6 +263,8 @@ void __init at91_add_device_usba(struct
 		usba_udc_data.pdata.vbus_pin = data->vbus_pin;
 	}
 
+	usba_udc_data.pdata.force_full_speed = data->force_full_speed;
+
 	/* Pullup pin is handled internally by USB device peripheral */
 
 	platform_device_register(&at91_usba_udc_device);
Index: linux-2.6.39/arch/arm/mach-at91/at91cap9_devices.c
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/at91cap9_devices.c
+++ linux-2.6.39/arch/arm/mach-at91/at91cap9_devices.c
@@ -179,6 +179,8 @@ void __init at91_add_device_usba(struct
 		usba_udc_data.pdata.vbus_pin = data->vbus_pin;
 	}
 
+	usba_udc_data.pdata.force_full_speed = data->force_full_speed;
+
 	/* Pullup pin is handled internally by USB device peripheral */
 
 	platform_device_register(&at91_usba_udc_device);
Index: linux-2.6.39/arch/arm/mach-at91/at91sam9rl_devices.c
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/at91sam9rl_devices.c
+++ linux-2.6.39/arch/arm/mach-at91/at91sam9rl_devices.c
@@ -153,6 +153,8 @@ void __init at91_add_device_usba(struct
 		usba_udc_data.pdata.vbus_pin = data->vbus_pin;
 	}
 
+	usba_udc_data.pdata.force_full_speed = data->force_full_speed;
+
 	/* Pullup pin is handled internally by USB device peripheral */
 
 	platform_device_register(&at91_usba_udc_device);
