Index: linux-2.6.39/drivers/net/smc91x.c
===================================================================
--- linux-2.6.39.orig/drivers/net/smc91x.c
+++ linux-2.6.39/drivers/net/smc91x.c
@@ -1948,7 +1948,10 @@ static int __devinit smc_probe(struct ne
 
 	/* Get the MAC address */
 	SMC_SELECT_BANK(lp, 1);
-	SMC_GET_MAC_ADDR(lp, dev->dev_addr);
+	if(is_valid_ether_addr(dev->dev_addr)) /* from pdata */
+		SMC_SET_MAC_ADDR(lp, dev->dev_addr);
+	else /* already set by bootloader ? */
+		SMC_GET_MAC_ADDR(lp, dev->dev_addr);
 
 	/* now, reset the chip, and put it into a known state */
 	smc_reset(dev);
@@ -2243,6 +2246,7 @@ static int __devinit smc_drv_probe(struc
 	if (pd) {
 		memcpy(&lp->cfg, pd, sizeof(lp->cfg));
 		lp->io_shift = SMC91X_IO_SHIFT(lp->cfg.flags);
+		memcpy(ndev->dev_addr, &pd->mac_addr, ETH_ALEN);
 	} else {
 		lp->cfg.flags |= (SMC_CAN_USE_8BIT)  ? SMC91X_USE_8BIT  : 0;
 		lp->cfg.flags |= (SMC_CAN_USE_16BIT) ? SMC91X_USE_16BIT : 0;
@@ -2299,6 +2303,7 @@ static int __devinit smc_drv_probe(struc
 		goto out_release_attrib;
 	}
 
+
 #ifdef CONFIG_ARCH_PXA
 	{
 		struct smc_local *lp = netdev_priv(ndev);
Index: linux-2.6.39/include/linux/smc91x.h
===================================================================
--- linux-2.6.39.orig/include/linux/smc91x.h
+++ linux-2.6.39/include/linux/smc91x.h
@@ -29,6 +29,7 @@ struct smc91x_platdata {
 	unsigned long flags;
 	unsigned char leda;
 	unsigned char ledb;
+	u8 mac_addr;
 };
 
 #endif /* __SMC91X_H__ */
