Index: linux-2.6.39/arch/arm/mach-at91/at91sam9g45_devices.c
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/at91sam9g45_devices.c
+++ linux-2.6.39/arch/arm/mach-at91/at91sam9g45_devices.c
@@ -1209,6 +1209,15 @@ void __init at91_add_device_pwm(u32 mask
 	if (mask & (1 << AT91_PWM3))
 		at91_set_B_periph(AT91_PIN_PD0, 1);	/* enable PWM3 */
 
+	if (mask & (1 << AT91_PWM1_ALT))
+		at91_set_B_periph(AT91_PIN_PD25, 1);	/* enable PWM1 alt. Pin */
+
+	if (mask & (1 << AT91_PWM2_ALT))
+		at91_set_A_periph(AT91_PIN_PE31, 1);	/* enable PWM2 alt. Pin*/
+
+	if (mask & (1 << AT91_PWM3_ALT))
+		at91_set_B_periph(AT91_PIN_PA25, 1);	/* enable PWM3 alt. Pin*/
+
 	pwm_mask = mask;
 
 	platform_device_register(&at91sam9g45_pwm0_device);
Index: linux-2.6.39/arch/arm/mach-at91/include/mach/board.h
===================================================================
--- linux-2.6.39.orig/arch/arm/mach-at91/include/mach/board.h
+++ linux-2.6.39/arch/arm/mach-at91/include/mach/board.h
@@ -163,10 +163,14 @@ extern void __init at91_add_device_seria
 /*
  * PWM
  */
-#define AT91_PWM0	0
-#define AT91_PWM1	1
-#define AT91_PWM2	2
-#define AT91_PWM3	3
+#define AT91_PWM0		0
+#define AT91_PWM1		1
+#define AT91_PWM2		2
+#define AT91_PWM3		3
+#define AT91_PWM0_ALT	4
+#define AT91_PWM1_ALT	5
+#define AT91_PWM2_ALT	6
+#define AT91_PWM3_ALT	7
 
 extern void __init at91_add_device_pwm(u32 mask);
 
