Index: ltp-full-20110915/testcases/kernel/syscalls/epoll/Makefile
===================================================================
--- ltp-full-20110915.orig/testcases/kernel/syscalls/epoll/Makefile
+++ ltp-full-20110915/testcases/kernel/syscalls/epoll/Makefile
@@ -20,4 +20,6 @@ top_srcdir		?= ../../../..
 
 include $(top_srcdir)/include/mk/testcases.mk
 
+CFLAGS			+= -O1
+
 include $(top_srcdir)/include/mk/generic_leaf_target.mk
