From f0c71adc287722427b8307e82eda2354bbd542b5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Uwe=20Kleine-K=C3=B6nig?= <u.kleine-koenig@pengutronix.de>
Date: Wed, 13 Jan 2010 17:58:04 +0100
Subject: [PATCH 31/59] imx: WARN in clk_disable if the clock isn't enabled
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@pengutronix.de>
Cc: Sascha Hauer <kernel@pengutronix.de>
Cc: Russell King <linux@arm.linux.org.uk>
---
 arch/arm/plat-mxc/clock.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

Index: linux-2.6.33/arch/arm/plat-mxc/clock.c
===================================================================
--- linux-2.6.33.orig/arch/arm/plat-mxc/clock.c
+++ linux-2.6.33/arch/arm/plat-mxc/clock.c
@@ -56,6 +56,7 @@ static void __clk_disable(struct clk *cl
 	__clk_disable(clk->parent);
 	__clk_disable(clk->secondary);
 
+	WARN_ON(!clk->usecount);
 	if (!(--clk->usecount) && clk->disable)
 		clk->disable(clk);
 }
