Index: linux-2.6.33/net/ipv4/ipconfig.c
===================================================================
--- linux-2.6.33.orig/net/ipv4/ipconfig.c
+++ linux-2.6.33/net/ipv4/ipconfig.c
@@ -1464,7 +1464,8 @@ static int __init ic_proto_name(char *na
 #endif
 #ifdef CONFIG_IP_PNP_BOOTP
 	else if (!strcmp(name, "bootp")) {
-		ic_proto_enabled &= ~(IC_RARP | IC_USE_DHCP);
+		/* For Garz & Fricke systems let bootp behave like dhcp */
+		ic_proto_enabled &= ~IC_RARP;
 		return 1;
 	}
 #endif
@@ -1544,9 +1545,20 @@ static int __init ip_auto_config_setup(c
 				strlcpy(user_dev_name, ip, sizeof(user_dev_name));
 				break;
 			case 6:
-				if (ic_proto_name(ip) == 0 &&
-				    ic_myaddr == NONE) {
-					ic_enable = 0;
+				if (ic_proto_name(ip) == 0) {
+				    if(ic_myaddr == NONE) {
+						ic_enable = 0;
+					} else {
+						ic_enable = 1;
+					}
+				 /* For Garz & Fricke systems discard all static IP stuff
+				    if bootp or dhcp */
+				} else {
+					ic_enable = 1;
+					ic_myaddr = NONE;
+					ic_servaddr = NONE;
+					ic_gateway = NONE;
+					ic_netmask = NONE;
 				}
 				break;
 			}
