Index: linux-2.6.33/arch/arm/mach-mx3/clock-imx35.c
===================================================================
--- linux-2.6.33.orig/arch/arm/mach-mx3/clock-imx35.c
+++ linux-2.6.33/arch/arm/mach-mx3/clock-imx35.c
@@ -496,10 +496,10 @@ int __init mx35_clocks_init()
 		clkdev_add(&lookups[i]);
 
 	/* Turn off all clocks except the ones we need to survive, namely:
-	 * EMI, GPIO1/2/3, GPT, IOMUX, MAX and eventually uart
+	 * EMI, GPIO1/2/3, GPT, IOMUX, MAX, IPU, PWM (backlight) and eventually uart
 	 */
 	__raw_writel((3 << 18), CCM_BASE + CCM_CGR0);
-	__raw_writel((3 << 2) | (3 << 4) | (3 << 6) | (3 << 8) | (3 << 16),
+	__raw_writel((3 << 2) | (3 << 4) | (3 << 6) | (3 << 8) | (3 << 16) | (3 << 18) | (3 << 28),
 			CCM_BASE + CCM_CGR1);
 	__raw_writel((3 << 26) | ll, CCM_BASE + CCM_CGR2);
 	__raw_writel(0, CCM_BASE + CCM_CGR3);
