Index: linux-2.6.33/arch/arm/mach-mx3/devices.c
===================================================================
--- linux-2.6.33.orig/arch/arm/mach-mx3/devices.c
+++ linux-2.6.33/arch/arm/mach-mx3/devices.c
@@ -595,6 +595,63 @@ struct platform_device imx_ssi_device1 =
 	.resource = imx_ssi_resources1,
 };
 
+#ifdef CONFIG_ARCH_MX35
+static struct resource mxc_sdhc_resources0[] = {
+	{
+		.start	= MX35_MMC_SDHC1_BASE_ADDR,
+		.end	= MX35_MMC_SDHC1_BASE_ADDR + SZ_4K - 1,
+		.flags	= IORESOURCE_MEM,
+	}, {
+		.start	= MX35_INT_MMC_SDHC1,
+		.end	= MX35_INT_MMC_SDHC1,
+		.flags	= IORESOURCE_IRQ,
+	},
+};
+
+struct platform_device mxc_sdhc_device0 = {
+	.name = "sdhci-esdhc-imx",
+	.id = 0,
+	.num_resources = ARRAY_SIZE(mxc_sdhc_resources0),
+	.resource = mxc_sdhc_resources0,
+};
+static struct resource mxc_sdhc_resources1[] = {
+	{
+		.start	= MX35_MMC_SDHC2_BASE_ADDR,
+		.end	= MX35_MMC_SDHC2_BASE_ADDR + SZ_4K - 1,
+		.flags	= IORESOURCE_MEM,
+	}, {
+		.start	= MX35_INT_MMC_SDHC2,
+		.end	= MX35_INT_MMC_SDHC2,
+		.flags	= IORESOURCE_IRQ,
+	},
+};
+
+struct platform_device mxc_sdhc_device1 = {
+	.name = "sdhci-esdhc-imx",
+	.id = 1,
+	.num_resources = ARRAY_SIZE(mxc_sdhc_resources1),
+	.resource = mxc_sdhc_resources1,
+};
+static struct resource mxc_sdhc_resources2[] = {
+	{
+		.start	= MX35_MMC_SDHC3_BASE_ADDR,
+		.end	= MX35_MMC_SDHC3_BASE_ADDR + SZ_4K - 1,
+		.flags	= IORESOURCE_MEM,
+	}, {
+		.start	= MX35_INT_MMC_SDHC3,
+		.end	= MX35_INT_MMC_SDHC3,
+		.flags	= IORESOURCE_IRQ,
+	},
+};
+
+struct platform_device mxc_sdhc_device2 = {
+	.name = "sdhci-esdhc-imx",
+	.id = 2,
+	.num_resources = ARRAY_SIZE(mxc_sdhc_resources2),
+	.resource = mxc_sdhc_resources2,
+};
+#endif
+
 static int mx3_devices_init(void)
 {
 	if (cpu_is_mx31()) {
Index: linux-2.6.33/arch/arm/mach-mx3/devices.h
===================================================================
--- linux-2.6.33.orig/arch/arm/mach-mx3/devices.h
+++ linux-2.6.33/arch/arm/mach-mx3/devices.h
@@ -26,4 +26,6 @@ extern struct platform_device mxc_spi_de
 extern struct platform_device mxc_spi_device2;
 extern struct platform_device imx_ssi_device0;
 extern struct platform_device imx_ssi_device1;
-
+extern struct platform_device mxc_sdhc_device0;
+extern struct platform_device mxc_sdhc_device1;
+extern struct platform_device mxc_sdhc_device2;
Index: linux-2.6.33/arch/arm/plat-mxc/include/mach/mx35.h
===================================================================
--- linux-2.6.33.orig/arch/arm/plat-mxc/include/mach/mx35.h
+++ linux-2.6.33/arch/arm/plat-mxc/include/mach/mx35.h
@@ -52,6 +52,9 @@
 #define MX35_GPIO3_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xa4000)
 #define MX35_SCC_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xac000)
 #define MX35_RNGA_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xb0000)
+#define MX35_MMC_SDHC1_BASE_ADDR		(MX35_AIPS2_BASE_ADDR + 0xb4000)
+#define MX35_MMC_SDHC2_BASE_ADDR		(MX35_AIPS2_BASE_ADDR + 0xb8000)
+#define MX35_MMC_SDHC3_BASE_ADDR		(MX35_AIPS2_BASE_ADDR + 0xbc000)
 #define MX35_IPU_CTRL_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xc0000)
 #define MX35_AUDMUX_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xc4000)
 #define MX35_GPIO1_BASE_ADDR			(MX35_AIPS2_BASE_ADDR + 0xcc000)
