Index: linux-2.6.33/drivers/mtd/nand/Kconfig
===================================================================
--- linux-2.6.33.orig/drivers/mtd/nand/Kconfig
+++ linux-2.6.33/drivers/mtd/nand/Kconfig
@@ -444,7 +444,7 @@ config MTD_NAND_FSL_UPM
 
 config MTD_NAND_MXC
 	tristate "MXC NAND support"
-	depends on ARCH_MX2 || ARCH_MX3
+	depends on ARCH_MX2 || ARCH_MX25 || ARCH_MX3
 	help
 	  This enables the driver for the NAND flash controller on the
 	  MXC processors.
Index: linux-2.6.33/drivers/video/Kconfig
===================================================================
--- linux-2.6.33.orig/drivers/video/Kconfig
+++ linux-2.6.33/drivers/video/Kconfig
@@ -402,7 +402,7 @@ config FB_SA1100
 
 config FB_IMX
 	tristate "Motorola i.MX LCD support"
-	depends on FB && (ARCH_MX1 || ARCH_MX2)
+	depends on FB && (ARCH_MX1 || ARCH_MX2 || ARCH_MX25)
 	select FB_CFB_FILLRECT
 	select FB_CFB_COPYAREA
 	select FB_CFB_IMAGEBLIT
