Index: linux-2.6.33/drivers/video/imxfb.c
===================================================================
--- linux-2.6.33.orig/drivers/video/imxfb.c
+++ linux-2.6.33/drivers/video/imxfb.c
@@ -326,7 +326,7 @@ static int imxfb_check_var(struct fb_var
 	struct imxfb_rgb *rgb;
 	const struct imx_fb_videomode *imxfb_mode;
 	unsigned long lcd_clk;
-	unsigned long long tmp;
+	unsigned long clk_div;
 	u32 pcr = 0;
 
 	if (var->xres < MIN_XRES)
@@ -356,16 +356,14 @@ static int imxfb_check_var(struct fb_var
 
 	lcd_clk = clk_get_rate(fbi->clk);
 
-	tmp = var->pixclock * (unsigned long long)lcd_clk;
+	if (lcd_clk % (var->pixclock * 1000) == 0)
+		clk_div = lcd_clk / (var->pixclock * 1000) - 1;
+	else
+		clk_div = lcd_clk / (var->pixclock * 1000);
 
-	do_div(tmp, 1000000);
+	pcr = (unsigned int)clk_div;
 
-	if (do_div(tmp, 1000000) > 500000)
-		tmp++;
-
-	pcr = (unsigned int)tmp;
-
-	if (--pcr > 0x3F) {
+	if (pcr > 0x3F) {
 		pcr = 0x3F;
 		printk(KERN_WARNING "Must limit pixel clock to %luHz\n",
 				lcd_clk / pcr);
