Index: linux-2.6.33/drivers/mtd/nand/mxc_nand.c
===================================================================
--- linux-2.6.33.orig/drivers/mtd/nand/mxc_nand.c
+++ linux-2.6.33/drivers/mtd/nand/mxc_nand.c
@@ -483,6 +483,17 @@ static void mxc_nand_select_chip(struct
 	}
 }
 
+/* This function overrides the default bad block table function.
+ * We do not want to have all bad blocks scanned at once, but
+ * only on access. Thus we do not create a bad block table at
+ * all, because the MTD layer creates its own one. */
+static int mxc_nand_bbt(struct mtd_info *mtd)
+{
+	struct nand_chip *this = mtd->priv;
+	this->bbt = NULL;
+	return 0;
+}
+
 /*
  * This function should keep manufacturer's bad block marker even in the case
  * the first generation NFC is in use. This controller can't handle 2kiB pages
@@ -811,6 +822,7 @@ static int __init mxcnd_probe(struct pla
 	this->write_buf = mxc_nand_write_buf;
 	this->read_buf = mxc_nand_read_buf;
 	this->verify_buf = mxc_nand_verify_buf;
+	this->scan_bbt = mxc_nand_bbt;
 
 	host->clk = clk_get(&pdev->dev, "nfc");
 	if (IS_ERR(host->clk)) {
