Index: linux-2.6.33/drivers/video/backlight/pwm_bl.c
===================================================================
--- linux-2.6.33.orig/drivers/video/backlight/pwm_bl.c
+++ linux-2.6.33/drivers/video/backlight/pwm_bl.c
@@ -26,6 +26,7 @@ struct pwm_bl_data {
 	unsigned int		period;
 	int			(*notify)(struct device *,
 					  int brightness);
+	uint16_t	*lut;
 };
 
 static int pwm_backlight_update_status(struct backlight_device *bl)
@@ -42,14 +43,25 @@ static int pwm_backlight_update_status(s
 
 	if (pb->notify)
 		brightness = pb->notify(pb->dev, brightness);
-
+#if 0
 	if (brightness == 0) {
-		pwm_config(pb->pwm, 0, pb->period);
+		if(pb->lut) {
+			printk("Setting backlight from LUT: Value: %d LUT Value %d\n", brightness, pb->lut[brightness]);
+			pwm_config_raw(pb->pwm, pb->lut[0], 0xFFFD);
+		} else {
+			pwm_config(pb->pwm, 0, pb->period);
+		}
 		pwm_disable(pb->pwm);
 	} else {
-		pwm_config(pb->pwm, brightness * pb->period / max, pb->period);
+#endif
+		if(pb->lut) {
+			pwm_config_raw(pb->pwm, pb->lut[brightness], 0xFFFD);
+		}
+		else {
+			pwm_config(pb->pwm, brightness * pb->period / max, pb->period);
+		}
 		pwm_enable(pb->pwm);
-	}
+//	}
 	return 0;
 }
 
@@ -91,6 +103,7 @@ static int pwm_backlight_probe(struct pl
 	pb->period = data->pwm_period_ns;
 	pb->notify = data->notify;
 	pb->dev = &pdev->dev;
+	pb->lut = data->lut;
 
 	pb->pwm = pwm_request(data->pwm_id, "backlight");
 	if (IS_ERR(pb->pwm)) {
@@ -132,7 +145,11 @@ static int pwm_backlight_remove(struct p
 	struct pwm_bl_data *pb = dev_get_drvdata(&bl->dev);
 
 	backlight_device_unregister(bl);
-	pwm_config(pb->pwm, 0, pb->period);
+	if(pb->lut) {
+		pwm_config_raw(pb->pwm, pb->lut[0], 0xFFFD);
+	} else {
+		pwm_config(pb->pwm, 0, pb->period);
+	}
 	pwm_disable(pb->pwm);
 	pwm_free(pb->pwm);
 	kfree(pb);
@@ -150,7 +167,11 @@ static int pwm_backlight_suspend(struct
 
 	if (pb->notify)
 		pb->notify(pb->dev, 0);
-	pwm_config(pb->pwm, 0, pb->period);
+	if(pb->lut) {
+		pwm_config_raw(pb->pwm, pb->lut[0], 0xFFFD);
+	} else {
+		pwm_config(pb->pwm, 0, 0xFFFD);
+	}
 	pwm_disable(pb->pwm);
 	return 0;
 }
Index: linux-2.6.33/include/linux/pwm_backlight.h
===================================================================
--- linux-2.6.33.orig/include/linux/pwm_backlight.h
+++ linux-2.6.33/include/linux/pwm_backlight.h
@@ -12,6 +12,7 @@ struct platform_pwm_backlight_data {
 	int (*init)(struct device *dev);
 	int (*notify)(struct device *dev, int brightness);
 	void (*exit)(struct device *dev);
+	uint16_t	*lut;
 };
 
 #endif
