Index: linux-2.6.33/drivers/usb/host/ehci-mxc.c
===================================================================
--- linux-2.6.33.orig/drivers/usb/host/ehci-mxc.c
+++ linux-2.6.33/drivers/usb/host/ehci-mxc.c
@@ -170,7 +170,7 @@ static int ehci_mxc_drv_probe(struct pla
 	}
 	clk_enable(priv->usbclk);
 
-	if (!cpu_is_mx35()) {
+	if (!cpu_is_mx35() && !cpu_is_mx25()) {
 		priv->ahbclk = clk_get(dev, "usb_ahb");
 		if (IS_ERR(priv->ahbclk)) {
 			ret = PTR_ERR(priv->ahbclk);
Index: linux-2.6.33/drivers/usb/gadget/fsl_mx3_udc.c
===================================================================
--- linux-2.6.33.orig/drivers/usb/gadget/fsl_mx3_udc.c
+++ linux-2.6.33/drivers/usb/gadget/fsl_mx3_udc.c
@@ -16,6 +16,7 @@
 #include <linux/err.h>
 #include <linux/fsl_devices.h>
 #include <linux/platform_device.h>
+#include <mach/hardware.h>
 
 static struct clk *mxc_ahb_clk;
 static struct clk *mxc_usb_clk;
@@ -28,14 +29,17 @@ int fsl_udc_clk_init(struct platform_dev
 
 	pdata = pdev->dev.platform_data;
 
-	mxc_ahb_clk = clk_get(&pdev->dev, "usb_ahb");
-	if (IS_ERR(mxc_ahb_clk))
-		return PTR_ERR(mxc_ahb_clk);
-
-	ret = clk_enable(mxc_ahb_clk);
-	if (ret < 0) {
-		dev_err(&pdev->dev, "clk_enable(\"usb_ahb\") failed\n");
-		goto eenahb;
+	if (!cpu_is_mx25())
+	{
+		mxc_ahb_clk = clk_get(&pdev->dev, "usb_ahb");
+		if (IS_ERR(mxc_ahb_clk))
+			return PTR_ERR(mxc_ahb_clk);
+
+		ret = clk_enable(mxc_ahb_clk);
+		if (ret < 0) {
+			dev_err(&pdev->dev, "clk_enable(\"usb_ahb\") failed\n");
+			goto eenahb;
+		}
 	}
 
 	/* make sure USB_CLK is running at 60 MHz +/- 1000 Hz */
@@ -66,9 +70,11 @@ eclkrate:
 	clk_put(mxc_usb_clk);
 	mxc_usb_clk = NULL;
 egusb:
-	clk_disable(mxc_ahb_clk);
+	if (!cpu_is_mx25())
+		clk_disable(mxc_ahb_clk);
 eenahb:
-	clk_put(mxc_ahb_clk);
+	if (!cpu_is_mx25())
+		clk_put(mxc_ahb_clk);
 	return ret;
 }
 
