Index: linux-2.6.33/sound/soc/imx/imx-pcm-fiq.c
===================================================================
--- linux-2.6.33.orig/sound/soc/imx/imx-pcm-fiq.c
+++ linux-2.6.33/sound/soc/imx/imx-pcm-fiq.c
@@ -32,6 +32,8 @@
 
 #include "imx-ssi.h"
 
+static void (*toggle_speaker)(int enable);
+
 struct imx_pcm_runtime_data {
 	int period;
 	int periods;
@@ -114,6 +116,10 @@ static int snd_imx_pcm_trigger(struct sn
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_RESUME:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+		/* Enable speaker */
+		if (toggle_speaker)
+			toggle_speaker(1);
+
 		iprtd->timer.expires = jiffies + iprtd->period_time;
 		add_timer(&iprtd->timer);
 		if (++fiq_enable == 1)
@@ -271,6 +277,9 @@ struct snd_soc_platform *imx_ssi_fiq_ini
 	ssi->dma_params_tx.burstsize = 4;
 	ssi->dma_params_rx.burstsize = 6;
 
+	/* Be able to enable/disable the speaker */
+	toggle_speaker = ssi->toggle_speaker;
+
 	return &imx_soc_platform_fiq;
 }
 
Index: linux-2.6.33/arch/arm/plat-mxc/include/mach/ssi.h
===================================================================
--- linux-2.6.33.orig/arch/arm/plat-mxc/include/mach/ssi.h
+++ linux-2.6.33/arch/arm/plat-mxc/include/mach/ssi.h
@@ -12,6 +12,7 @@ struct imx_ssi_platform_data {
 #define IMX_SSI_USE_AC97       (1 << 1)
 	void (*ac97_reset) (struct snd_ac97 *ac97);
 	void (*ac97_warm_reset)(struct snd_ac97 *ac97);
+	void (*toggle_speaker)(int enable);
 };
 
 #endif /* __MACH_SSI_H */
Index: linux-2.6.33/sound/soc/imx/imx-ssi.c
===================================================================
--- linux-2.6.33.orig/sound/soc/imx/imx-ssi.c
+++ linux-2.6.33/sound/soc/imx/imx-ssi.c
@@ -807,6 +807,7 @@ static int imx_ssi_probe(struct platform
 	if (pdata) {
 		ssi->ac97_reset = pdata->ac97_reset;
 		ssi->ac97_warm_reset = pdata->ac97_warm_reset;
+		ssi->toggle_speaker = pdata->toggle_speaker;
 		ssi->flags = pdata->flags;
 	}
 
Index: linux-2.6.33/sound/soc/imx/imx-ssi.h
===================================================================
--- linux-2.6.33.orig/sound/soc/imx/imx-ssi.h
+++ linux-2.6.33/sound/soc/imx/imx-ssi.h
@@ -215,6 +215,8 @@ struct imx_ssi {
 	void (*ac97_reset) (struct snd_ac97 *ac97);
 	void (*ac97_warm_reset)(struct snd_ac97 *ac97);
 
+	void (*toggle_speaker)(int enable);
+
 	struct imx_pcm_dma_params	dma_params_rx;
 	struct imx_pcm_dma_params	dma_params_tx;
 
Index: linux-2.6.33/sound/soc/imx/imx-pcm-dma-mx2.c
===================================================================
--- linux-2.6.33.orig/sound/soc/imx/imx-pcm-dma-mx2.c
+++ linux-2.6.33/sound/soc/imx/imx-pcm-dma-mx2.c
@@ -30,6 +30,8 @@
 
 #include "imx-ssi.h"
 
+static void (*toggle_speaker)(int enable);
+
 struct imx_pcm_runtime_data {
 	int sg_count;
 	struct scatterlist *sg_list;
@@ -228,14 +230,18 @@ static int snd_imx_pcm_trigger(struct sn
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_RESUME:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
-		imx_dma_enable(iprtd->dma);
+		/* Enable speaker */
+		if (toggle_speaker)
+			toggle_speaker(1);
+
+			imx_dma_enable(iprtd->dma);
 
 		break;
 
 	case SNDRV_PCM_TRIGGER_STOP:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		imx_dma_disable(iprtd->dma);
+			imx_dma_disable(iprtd->dma);
 
 		break;
 	default:
@@ -314,6 +320,9 @@ struct snd_soc_platform *imx_ssi_dma_mx2
 	ssi->dma_params_tx.burstsize = DMA_TXFIFO_BURST;
 	ssi->dma_params_rx.burstsize = DMA_RXFIFO_BURST;
 
+	/* Be able to enable/disable the speaker */
+	toggle_speaker = ssi->toggle_speaker;
+
 	return &imx_soc_platform_dma;
 }
 
