Index: linux-2.6.33/drivers/i2c/busses/i2c-imx.c
===================================================================
--- linux-2.6.33.orig/drivers/i2c/busses/i2c-imx.c
+++ linux-2.6.33/drivers/i2c/busses/i2c-imx.c
@@ -521,6 +521,7 @@ static int __init i2c_imx_probe(struct p
 	strcpy(i2c_imx->adapter.name, pdev->name);
 	i2c_imx->adapter.owner		= THIS_MODULE;
 	i2c_imx->adapter.algo		= &i2c_imx_algo;
+	i2c_imx->adapter.class = I2C_CLASS_HWMON;
 	i2c_imx->adapter.dev.parent	= &pdev->dev;
 	i2c_imx->adapter.nr 		= pdev->id;
 	i2c_imx->irq			= irq;
