Index: linux-2.6.33/drivers/char/vt.c
===================================================================
--- linux-2.6.33.orig/drivers/char/vt.c
+++ linux-2.6.33/drivers/char/vt.c
@@ -177,8 +177,12 @@ int do_poke_blanked_console;
 int console_blanked;
 
 static int vesa_blank_mode; /* 0:none 1:suspendV 2:suspendH 3:powerdown */
+#ifndef CONFIG_FB_PRE_INIT_FB
+static int blankinterval = 10*60*HZ;
+#else
+static int blankinterval = 0;
+#endif
 static int vesa_off_interval;
-static int blankinterval = 10*60;
 core_param(consoleblank, blankinterval, int, 0444);
 
 static DECLARE_WORK(console_work, console_callback);
@@ -698,8 +702,10 @@ void redraw_screen(struct vc_data *vc, i
 			update_attr(vc);
 			clear_buffer_attributes(vc);
 		}
+#ifndef CONFIG_FB_PRE_INIT_FB
 		if (update && vc->vc_mode != KD_GRAPHICS)
 			do_update_region(vc, vc->vc_origin, vc->vc_screenbuf_size / 2);
+#endif
 	}
 	set_cursor(vc);
 	if (is_switch) {
@@ -783,8 +789,11 @@ int vc_allocate(unsigned int currcons)	/
 	       boot option, default to displaying the cursor */
 	    if (global_cursor_default == -1)
 		    global_cursor_default = 1;
-
+#ifdef CONFIG_FB_PRE_INIT_FB
+	    vc_init(vc, vc->vc_rows, vc->vc_cols, 0);
+#else
 	    vc_init(vc, vc->vc_rows, vc->vc_cols, 1);
+#endif
 	    vcs_make_sysfs(currcons);
 	    atomic_notifier_call_chain(&vt_notifier_list, VT_ALLOCATE, &param);
 	}
Index: linux-2.6.33/drivers/video/Kconfig
===================================================================
--- linux-2.6.33.orig/drivers/video/Kconfig
+++ linux-2.6.33/drivers/video/Kconfig
@@ -406,6 +406,7 @@ config FB_IMX
 	select FB_CFB_FILLRECT
 	select FB_CFB_COPYAREA
 	select FB_CFB_IMAGEBLIT
+	select FB_PRE_INIT_FB
 
 config FB_CYBER2000
 	tristate "CyberPro 2000/2010/5000 support"
@@ -2124,6 +2125,7 @@ config FB_MB862XX_LIME
 	depends on PPC
 	select FB_FOREIGN_ENDIAN
 	select FB_LITTLE_ENDIAN
+	select FB_PRE_INIT_FB
 	---help---
 	  Framebuffer support for Fujitsu Lime GDC on host CPU bus.
 
@@ -2139,11 +2141,7 @@ config FB_EP93XX
 	  ep93xx-fb.
 
 config FB_PRE_INIT_FB
-	bool "Don't reinitialize, use bootloader's GDC/Display configuration"
-	depends on FB && FB_MB862XX_LIME
-	---help---
-	  Select this option if display contents should be inherited as set by
-	  the bootloader.
+	bool
 
 config FB_MSM
 	tristate "MSM Framebuffer support"
@@ -2159,6 +2157,7 @@ config FB_MX3
 	select FB_CFB_FILLRECT
 	select FB_CFB_COPYAREA
 	select FB_CFB_IMAGEBLIT
+	select FB_PRE_INIT_FB
 	default y
 	help
 	  This is a framebuffer device for the i.MX31 LCD Controller. So
Index: linux-2.6.33/drivers/video/console/fbcon.c
===================================================================
--- linux-2.6.33.orig/drivers/video/console/fbcon.c
+++ linux-2.6.33/drivers/video/console/fbcon.c
@@ -1073,6 +1073,9 @@ static void fbcon_init(struct vc_data *v
 	if (p->userfont)
 		charcnt = FNTCHARCNT(p->fontdata);
 
+#if defined(CONFIG_FB_PRE_INIT_FB)
+       vc->vc_deccm = 0;
+#endif
 	vc->vc_can_do_color = (fb_get_color_depth(&info->var, &info->fix)!=1);
 	vc->vc_complement_mask = vc->vc_can_do_color ? 0x7700 : 0x0800;
 	if (charcnt == 256) {
@@ -3071,7 +3074,11 @@ static int fbcon_fb_registered(struct fb
 		}
 
 		if (info_idx != -1)
+#ifdef CONFIG_FB_PRE_INIT_FB
+			ret = fbcon_takeover(0);
+#else
 			ret = fbcon_takeover(1);
+#endif
 	} else {
 		for (i = first_fb_vc; i <= last_fb_vc; i++) {
 			if (con2fb_map_boot[i] == idx)
