Index: linux-2.6.33/drivers/video/console/fbcon.c
===================================================================
--- linux-2.6.33.orig/drivers/video/console/fbcon.c
+++ linux-2.6.33/drivers/video/console/fbcon.c
@@ -502,6 +502,29 @@ static int __init fb_console_setup(char
 __setup("fbcon=", fb_console_setup);
 #endif
 
+void fbcon_change_rotation(int rotation)
+{
+	struct fb_info *info;
+	int rotate, idx;
+
+	if (fbcon_has_exited)
+		return;
+
+	acquire_console_sem();
+	idx = con2fb_map[fg_console];
+
+	if (idx == -1 || registered_fb[idx] == NULL)
+		goto err;
+
+	info = registered_fb[idx];
+	rotate = rotation;
+	fbcon_rotate(info, rotate);
+err:
+	release_console_sem();
+	return;
+}
+EXPORT_SYMBOL_GPL(fbcon_change_rotation);
+
 static int search_fb_in_map(int idx)
 {
 	int i, retval = 0;
Index: linux-2.6.33/include/linux/fb.h
===================================================================
--- linux-2.6.33.orig/include/linux/fb.h
+++ linux-2.6.33/include/linux/fb.h
@@ -953,6 +953,9 @@ extern ssize_t fb_sys_read(struct fb_inf
 extern ssize_t fb_sys_write(struct fb_info *info, const char __user *buf,
 			    size_t count, loff_t *ppos);
 
+/* drivers/video/fbcon.c */
+extern void fbcon_change_rotation(int rotation);
+
 /* drivers/video/fbmem.c */
 extern int register_framebuffer(struct fb_info *fb_info);
 extern int unregister_framebuffer(struct fb_info *fb_info);
