Index: linux-2.6.33/drivers/video/backlight/backlight.c
===================================================================
--- linux-2.6.33.orig/drivers/video/backlight/backlight.c
+++ linux-2.6.33/drivers/video/backlight/backlight.c
@@ -13,65 +13,42 @@
 #include <linux/ctype.h>
 #include <linux/err.h>
 #include <linux/fb.h>
+#include <linux/platform_device.h>
 
 #ifdef CONFIG_PMAC_BACKLIGHT
 #include <asm/backlight.h>
 #endif
 
-#if defined(CONFIG_FB) || (defined(CONFIG_FB_MODULE) && \
-			   defined(CONFIG_BACKLIGHT_CLASS_DEVICE_MODULE))
-/* This callback gets called when something important happens inside a
- * framebuffer driver. We're looking if that important event is blanking,
- * and if it is, we're switching backlight power as well ...
- */
-static int fb_notifier_callback(struct notifier_block *self,
-				unsigned long event, void *data)
+unsigned long backlight_get_power(struct device *dev)
 {
-	struct backlight_device *bd;
-	struct fb_event *evdata = data;
-
-	/* If we aren't interested in this event, skip it immediately ... */
-	if (event != FB_EVENT_BLANK && event != FB_EVENT_CONBLANK)
-		return 0;
-
-	bd = container_of(self, struct backlight_device, fb_notif);
-	mutex_lock(&bd->ops_lock);
-	if (bd->ops)
-		if (!bd->ops->check_fb ||
-		    bd->ops->check_fb(evdata->info)) {
-			bd->props.fb_blank = *(int *)evdata->data;
-			if (bd->props.fb_blank == FB_BLANK_UNBLANK)
-				bd->props.state &= ~BL_CORE_FBBLANK;
-			else
-				bd->props.state |= BL_CORE_FBBLANK;
-			backlight_update_status(bd);
-		}
-	mutex_unlock(&bd->ops_lock);
-	return 0;
+	struct platform_device *pd = to_platform_device(dev);
+	struct backlight_device *bd = platform_get_drvdata(pd);
+	
+	if (bd)
+		return bd->props.power;
+	else
+		return ENODATA;
 }
 
-static int backlight_register_fb(struct backlight_device *bd)
-{
-	memset(&bd->fb_notif, 0, sizeof(bd->fb_notif));
-	bd->fb_notif.notifier_call = fb_notifier_callback;
-
-	return fb_register_client(&bd->fb_notif);
-}
+EXPORT_SYMBOL(backlight_get_power);
 
-static void backlight_unregister_fb(struct backlight_device *bd)
-{
-	fb_unregister_client(&bd->fb_notif);
-}
-#else
-static inline int backlight_register_fb(struct backlight_device *bd)
+int backlight_set_power(struct device *dev, unsigned long power)
 {
-	return 0;
-}
+	struct platform_device *pd = to_platform_device(dev);
+	struct backlight_device *bd = platform_get_drvdata(pd);
 
-static inline void backlight_unregister_fb(struct backlight_device *bd)
-{
+	if (bd)
+	{
+		pr_debug("backlight: set power to %lu\n", power);
+		bd->props.power = power;
+		backlight_update_status(bd);
+		return 0;
+	}
+	else
+		return ENODATA;
 }
-#endif /* CONFIG_FB */
+
+EXPORT_SYMBOL(backlight_set_power);
 
 static void backlight_generate_event(struct backlight_device *bd,
 				     enum backlight_update_reason reason)
@@ -295,12 +272,6 @@ struct backlight_device *backlight_devic
 		return ERR_PTR(rc);
 	}
 
-	rc = backlight_register_fb(new_bd);
-	if (rc) {
-		device_unregister(&new_bd->dev);
-		return ERR_PTR(rc);
-	}
-
 	new_bd->ops = ops;
 
 #ifdef CONFIG_PMAC_BACKLIGHT
@@ -335,7 +306,6 @@ void backlight_device_unregister(struct
 	bd->ops = NULL;
 	mutex_unlock(&bd->ops_lock);
 
-	backlight_unregister_fb(bd);
 	device_unregister(&bd->dev);
 }
 EXPORT_SYMBOL(backlight_device_unregister);
Index: linux-2.6.33/include/linux/backlight.h
===================================================================
--- linux-2.6.33.orig/include/linux/backlight.h
+++ linux-2.6.33/include/linux/backlight.h
@@ -108,6 +108,10 @@ extern void backlight_device_unregister(
 extern void backlight_force_update(struct backlight_device *bd,
 				   enum backlight_update_reason reason);
 
+
+extern unsigned long backlight_get_power(struct device *dev);
+extern int backlight_set_power(struct device *dev, unsigned long power);
+
 #define to_backlight_device(obj) container_of(obj, struct backlight_device, dev)
 
 static inline void * bl_get_data(struct backlight_device *bl_dev)
