Index: linux-2.6.33/init/main.c
===================================================================
--- linux-2.6.33.orig/init/main.c	2011-09-27 14:09:42.701634659 +0200
+++ linux-2.6.33/init/main.c	2011-09-27 14:09:47.844220000 +0200
@@ -238,8 +238,15 @@
 	return 0;
 }
 
+static int __init super_quiet_kernel(char *str)
+{
+	console_loglevel = 0;
+	return 0;
+}
+
 early_param("debug", debug_kernel);
 early_param("quiet", quiet_kernel);
+early_param("-q", super_quiet_kernel);
 
 static int __init loglevel(char *str)
 {
