Index: linux-2.6.33/drivers/base/firmware_class.c
===================================================================
--- linux-2.6.33.orig/drivers/base/firmware_class.c
+++ linux-2.6.33/drivers/base/firmware_class.c
@@ -33,7 +33,7 @@ enum {
 	FW_STATUS_ABORT,
 };
 
-static int loading_timeout = 60;	/* In seconds */
+static int loading_timeout = 1;	/* In seconds */
 
 /* fw_lock could be moved to 'struct firmware_priv' but since it is just
  * guarding for corner cases a global lock should be OK */
