Index: linux-2.6.33/scripts/setlocalversion
===================================================================
--- linux-2.6.33.orig/scripts/setlocalversion
+++ linux-2.6.33/scripts/setlocalversion
@@ -72,9 +72,16 @@ if hgid=`hg id 2>/dev/null`; then
 	exit
 fi
 
+# Check for local svn version file.
+if rev=`cat ../../../.svnrev 2>/dev/null`; then
+	printf -- '-svn%s' "$rev"
+
+	# All done with local svn version file
+	exit
+fi
+
 # Check for svn and a svn repo.
-if rev=`svn info 2>/dev/null | grep '^Last Changed Rev'`; then
-	rev=`echo $rev | awk '{print $NF}'`
+if rev=`cd patches; svnversion 2>/dev/null | grep -v "Unversioned"`; then
 	printf -- '-svn%s' "$rev"
 
 	# All done with svn
