Index: linux-2.6.33/arch/arm/kernel/setup.c
===================================================================
--- linux-2.6.33.orig/arch/arm/kernel/setup.c
+++ linux-2.6.33/arch/arm/kernel/setup.c
@@ -78,6 +78,9 @@ unsigned int __atags_pointer __initdata;
 unsigned int system_rev;
 EXPORT_SYMBOL(system_rev);
 
+unsigned int redboot_rev;
+EXPORT_SYMBOL(redboot_rev);
+
 unsigned int system_serial_low;
 EXPORT_SYMBOL(system_serial_low);
 
@@ -623,6 +626,7 @@ __tagtable(ATAG_SERIAL, parse_tag_serial
 static int __init parse_tag_revision(const struct tag *tag)
 {
 	system_rev = tag->u.revision.rev;
+	redboot_rev = tag->u.revision.rev;
 	return 0;
 }
 
@@ -865,6 +869,12 @@ static int c_show(struct seq_file *m, vo
 	seq_printf(m, "Serial\t\t: %08x%08x\n",
 		   system_serial_high, system_serial_low);
 
+	seq_puts(m, "\n");
+	seq_printf(m, "RedBoot\t\t: %u.%ur%u\n",
+		((redboot_rev >> 24) & 0xff),
+		((redboot_rev >> 16) & 0xff),
+		(redboot_rev & 0xffff));
+
 	return 0;
 }
 
Index: linux-2.6.33/include/linux/guf_board_common.h
===================================================================
--- /dev/null
+++ linux-2.6.33/include/linux/guf_board_common.h
@@ -0,0 +1,26 @@
+/*
+ * Common code for Garz & Fricke boards
+ *
+ * Copyright 2012 Garz & Fricke GmbH
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ */
+
+#ifndef GUF_BOARD_COMMON_H
+#define GUF_BOARD_COMMON_H
+
+int guf_check_redboot_compatibility(char* supported_redboot_version);
+
+#endif // GUF_BOARD_COMMON_H
Index: linux-2.6.33/lib/guf_board_common.c
===================================================================
--- /dev/null
+++ linux-2.6.33/lib/guf_board_common.c
@@ -0,0 +1,56 @@
+/*
+ * Common code for Garz & Fricke boards
+ *
+ * Copyright 2012 Garz & Fricke GmbH
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ */
+
+#include <linux/kernel.h>
+#include <linux/errno.h>
+
+extern unsigned int redboot_rev;
+
+int guf_check_redboot_compatibility(char* supported_redboot_version)
+{
+	int supported_major, supported_minor, supported_svn;
+	int current_major, current_minor, current_svn;
+	int ret;
+
+	ret = sscanf(supported_redboot_version, "%u.%ur%u",
+		&supported_major, &supported_minor, &supported_svn);
+	if (ret != 3) {
+		printk("ERROR: Could not determine supported RedBoot version\n");
+		return -EINVAL;
+	}
+
+	current_major = (redboot_rev >> 24) & 0xff;
+	current_minor = (redboot_rev >> 16) & 0xff;
+	current_svn = redboot_rev & 0xffff;
+
+	if ((current_major < supported_major) ||
+		(current_major == supported_major && current_minor < supported_minor) ||
+		(current_svn < supported_svn))
+	{
+		printk("ERROR: Current RedBoot version is not supported by this kernel.\n");
+		printk("Minimum required RedBoot version: %u.%ur%u\n",
+								supported_major, supported_minor, supported_svn);
+		printk("Current RedBoot version: %u.%ur%u\n",
+								current_major, current_minor, current_svn);
+		return -EPERM;
+	}
+
+	return 0;
+}
Index: linux-2.6.33/lib/Makefile
===================================================================
--- linux-2.6.33.orig/lib/Makefile
+++ linux-2.6.33/lib/Makefile
@@ -17,7 +17,7 @@ lib-y := ctype.o string.o vsprintf.o cmd
 lib-$(CONFIG_MMU) += ioremap.o
 lib-$(CONFIG_SMP) += cpumask.o
 
-lib-y	+= kobject.o kref.o klist.o
+lib-y	+= kobject.o kref.o klist.o guf_board_common.o
 
 obj-y += bcd.o div64.o sort.o parser.o halfmd4.o debug_locks.o random32.o \
 	 bust_spinlocks.o hexdump.o kasprintf.o bitmap.o scatterlist.o \
