Index: linux-2.6.33/arch/arm/mach-mx3/mach-guf_cupid.c
===================================================================
--- linux-2.6.33.orig/arch/arm/mach-mx3/mach-guf_cupid.c
+++ linux-2.6.33/arch/arm/mach-mx3/mach-guf_cupid.c
@@ -286,8 +286,6 @@ static void guf_cupid_setup_fb_videomode
 {
 	guf_cupid_mx3fb_pdata.name = config.display.name;
 	guf_cupid_mx3fb_pdata.bpp = config.format.depth;
-	guf_cupid_fb_mode.xres = config.display.xres;
-	guf_cupid_fb_mode.yres = config.display.yres;
 	guf_cupid_fb_mode.pixclock = (config.display.xres + config.hsync.start_width \
 			+ config.hsync.width + config.hsync.end_width) * \
 			(config.display.yres + config.vsync.start_width + config.vsync.width \
@@ -328,6 +326,31 @@ static void guf_cupid_setup_fb_videomode
 	guf_cupid_mx3fb_pdata.dma_dev = &mx3_ipu.dev;
 	guf_cupid_mx3fb_pdata.num_modes = 1;
 
+	switch (config.rotation.angle)
+	{
+		default:
+		case ROTATION_0:
+			guf_cupid_fb_mode.xres = config.display.xres;
+			guf_cupid_fb_mode.yres = config.display.yres;
+			guf_cupid_mx3fb_pdata.rotation = IPU_ROTATE_NONE;
+			break;
+		case ROTATION_90:
+			guf_cupid_fb_mode.xres = config.display.yres;
+			guf_cupid_fb_mode.yres = config.display.xres;
+			guf_cupid_mx3fb_pdata.rotation = IPU_ROTATE_90_RIGHT;
+			break;
+		case ROTATION_180:
+			guf_cupid_fb_mode.xres = config.display.xres;
+			guf_cupid_fb_mode.yres = config.display.yres;
+			guf_cupid_mx3fb_pdata.rotation = IPU_ROTATE_180;
+			break;
+		case ROTATION_270:
+			guf_cupid_fb_mode.xres = config.display.yres;
+			guf_cupid_fb_mode.yres = config.display.xres;
+			guf_cupid_mx3fb_pdata.rotation = IPU_ROTATE_90_LEFT;
+			break;
+	}
+
 	/* If the bootloader provides a framebuffer, we will continue to use it */
 	if (fixed_screen_address != 0U) {
 		guf_cupid_mx3fb_pdata.fixed_screen_cpu = phys_to_virt(fixed_screen_address);
